id/id.1
.\"
.\" Copyright (c) 2003 Gunnar Ritter
.\"
.\" SPDX-Licence-Identifier: Zlib
.\"
.\" Sccsid @(#)id.1	1.10 (gritter) 1/24/05
.TH ID 1 "1/24/05" "Heirloom Toolchest" "User Commands"
.SH NAME
id \- print user and group IDs and names
.SH SYNOPSIS
\fB/usr/5bin/id\fR
.br
\fB/usr/5bin/id\fR \fB\-a\fR
.sp
\fB/usr/5bin/posix/id\fR [\fIuser\fR]
.br
\fB/usr/5bin/posix/id\fR \fB\-a\fR [\fIuser\fR]
.br
\fB/usr/5bin/posix/id\fR \fB\-G\fR [\fB\-n\fR] [\fIuser\fR]
.br
\fB/usr/5bin/posix/id\fR \fB\-g\fR [\fB\-nr\fR] [\fIuser\fR]
.br
\fB/usr/5bin/posix/id\fR \fB\-u\fR [\fB\-nr\fR] [\fIuser\fR]
.SH DESCRIPTION
The
.I id
command outputs the numerical and symbolic
user and group IDs
of the invoking user or,
if the
.I user
argument is given for
.BR /usr/5bin/posix/id ,
for the named
.IR user.
If an effective ID does not match the corresponding real ID,
both are printed.
.PP
The
.B /usr/5bin/posix/id
adds supplementary group memberships to the output by default.
.PP
.B /usr/5bin/s42/id
and
.B /usr/5bin/posix2001/id
are identical to
.BR /usr/5bin/posix/id .
.PP
The following options are accepted:
.SS /usr/5bin/id
.TP
.B \-a
Adds supplementary group memberships to the output.
.SS /usr/5bin/posix/id
.TP
.B \-a
Adds the primary group to the list of supplementary groups.
.TP
.B \-G
Writes real, effective and supplementary group ID numbers only.
.TP
.B \-g
Writes the effective group ID number only.
.TP
.B \-n
Modifies the output of the
.BR \-G ,
.BR \-g ,
and
.B \-u
options to print symbolic instead of numeric IDs.
.TP
.B \-r
Causes real instead of effective IDs to be printed.
.TP
.B \-u
Writes the effective user ID number only.
.SH FILES
.TP
.B /etc/passwd
.TP
.B /etc/group
.SH "SEE ALSO"
groups(1),
logins(1),
getuid(2),
geteuid(2),
getgroups(2),
group(5),
passwd(5)
.SH NOTES
Use of
.I id
with any options or arguments is unportable.
For portable programs, use
.IR sed (1)
to cut the appropriate portions off the default output of
.I id
(using the invariable `uid=' and `gid=' strings as orientation)
and
.IR groups (1)
or
.IR logins (1)
if you need more information than available by this method.


CHANGES
Release 070715
* cpio: A "-P" option has been introduced that allows to specify attributes
  like the file owner when creating an archive. With it, unprivileged users
  can set arbitrary attributes for files within an archive.
* diff: Made it compile on CentOS 5.
* Static linking against glibc is now possible since "optopt" is no longer
  a common symbol, and there is no conflict anymore with "getopt.o".
* cpio, tar: A compilation problem on some Linux systems has been fixed.
* ul: A compilation problem for NetBSD and other systems that use termcap
  instead of curses has been fixed (patch by Aleksey Cheusov).
* bfs: An OpenSolaris-derived manual page is now included.

Release 070227
* The directory for development utilities, by default "/usr/ccs/bin", is
  now configurable in "build/mk.config" . It is expected that the utilities
  from the "Heirloom Development Tools" package are installed in it.
* getconf: "/usr/ccs/bin" is now included in the default path as printed
  with the "PATH" argument.
* what: Has been replaced by an OpenSolaris-derived implementation.
* tsort: Is now installed in "/usr/ccs/bin".
* OpenSolaris-derived manual pages are now included for bdiff, newform,
  and fspec(5).
* psrinfo: Fixed the computation of the "on-line since" time for Linux 2.6.
* cpio: Added a "-D" option to make it usable in combination with SVR4
  packaging tools. This option is not intended for general use, and is
  thus not documented.
* The package build target "make pkg" for Solaris and Open UNIX is no
  longer supported.

Release 070115
* deroff: Now recognizes the ".do" and ".xflag" requests as well as an "-xN"
  command line option to handle Heirloom nroff/troff extensions.
* ed: Substituting only a specific occurence of the pattern, as in the
  command "s/x/y/3", now works correctly if the substitution "y" is not
  equal in length to the pattern match "x" (Derek Thomas).
* ps: Use the Linux 2.6 /proc/status "SleepAVG" field to compute the value
  for the "C" column (cpu usage for scheduling).
* priocntl: The "FF", "B", and "ISO" class names are now accepted with
  the "-e" option too.
* file: UTF-8 characters in troff input files are now recognized.
* cp: The "-s" option now displays correct sizes and related statistics
  for special files with st_size == 0.
* grep: With the "-z" option, decompressor children processes are now
  properly waited for, avoiding failures when processing many files.
* su: Works with PAM on Mac OS X now (Stefan Tramm).
* ps: Handles dead processes as on other platforms now (Stefan Tramm).
* pgrep: Ignores dead processes now (Stefan Tramm).
* spell: A value of the wrong type that was passed to scanf() could lead
  to build errors on 64-bit platforms (Jeffrey H. Johnson).
* ps, pgrep, whodo: Works on FreeBSD 6 now.
* df, find, ps: Changes have been made to build on NetBSD 3 (Aleksey Cheusov).
* Changes have been made to build on OpenBSD 3.8 (Aleksey Cheusov).
* Changes have been made to build with GNU make 3.81.

Release 060122
* The tools have been ported to Mac OS X 10.4.3 (patches by Stefan Tramm).
* The tools have been ported to DragonFlyBSD 1.3.7-DEVELOPMENT.

Release 051118
* rm: The output of the question before descending to a directory with
  the "-i" option has been fixed.
* /usr/ucb/ps: The "-a" option has been fixed; it incorrectly printed a
  list of all processes since release 050411.
* getconf: /usr/ucb is now put at the end of the Heirloom directory list
  for the PATH variable.
* The "-man" macro no longer produces cut marks when used with troff.
* tr: When evaluating '[:class:]' expressions, the last position of the
  coded character set was not included; this has been fixed.
* ps: The "-R" option (resource usage format), which had been present
  for some time, has been documented.
* RPM package builds now require the new "Heirloom Development Tools"
  package from <http://heirloom.sourceforge.net/devtools.html> to be
  installed.

Release 050815
* The "getconf" utility has been added (derived from OpenSolaris code).
* The "mesg" and "pathchk" utilities have been added.
* tr: A new implementation that is based on 6th and 7th edition code is
  now used. It stores the translation tables in arrays of variable size.
  This results in better performance, under some circumstances at the
  expense of higher memory consumption.
* printf: Handle unsigned values larger than INT_MAX correctly with "%u",
  "%o", "%x", and "%X".
* printf: Handle SVID3 "position" fields in format specifications, such
  as "%7$s", that specify which argument is used next.
* pg: If a line does not fit on the screen, do not print a newline to
  force wraparound unless it is really necessary. This allows to select
  words across the wrap for copy-and-paste in an xterm.
* pg: Discard the ESC[48m sequence on a terminal that supports colors.
* ed: Do not print "expecting `w'" with the "q" command after a zero-byte
  file has been written.
* man: The tmac.an macros no longer use "@" as an escape character; its
  previous use lead to weird output with Unix nroff when an "@" appeared
  in titles or hanging labels.
* man: The "TCAT" post processor is now invoked separately for each manual
  entry. Previously, when the -t option was given with multiple entries and
  TCAT=dpost, only the first entry was actually printed.
* Some issues with building on AIX have been fixed.

Release 050713
* The 'echo' and 'test' commands have been added as standalone versions of
  those of the Heirloom Bourne shell.
* The 'false', 'kill', 'printf', 'pwd', and 'true' commands have been added.
* ed: Copying a larger number of lines with the 't' command could cause a
  segmentation fault because some pointers were not updated after calling
  realloc().
* ed: Do not warn about an 'input error' anymore when reading a file that
  lacks the terminating newline.
* ed: Disallow the 'u' command if no change has been made yet after a file
  specified on the command line has been read.

id/id.c
/*
 * id - print user and group IDs and names
 *
 * Gunnar Ritter, Freiburg i. Br., Germany, August 2002.
 */
/*
 * Copyright (c) 2003 Gunnar Ritter
 *
 * SPDX-Licence-Identifier: Zlib
 */

#if __GNUC__ >= 3 && __GNUC_MINOR__ >= 4 || __GNUC__ >= 4
#define	USED	__attribute__ ((used))
#elif defined __GNUC__
#define	USED	__attribute__ ((unused))
#else
#define	USED
#endif
#ifdef	SUS
static const char sccsid[] USED = "@(#)id_sus.sl	1.10 (gritter) 5/29/05";
#else
static const char sccsid[] USED = "@(#)id.sl	1.10 (gritter) 5/29/05";
#endif

#include	<unistd.h>
#include	<stdio.h>
#include	<string.h>
#include	<stdlib.h>
#include	<errno.h>
#include	<libgen.h>
#include	<pwd.h>
#include	<grp.h>

static int	aflag;			/* print supplementary GIDs */
static int	nflag;			/* print names instead of numbers */
static int	rflag;			/* use real IDs */
static int	restriction;		/* 'G' or 'g' or 'u' */
static char	*progname;		/* argv[0] to main() */

static void *
srealloc(void *vp, size_t nbytes)
{
	void	*p;

	if ((p = (void *)realloc(vp, nbytes)) == NULL) {
		write(2, "Out of memory\n", 14);
		exit(077);
	}
	return p;
}

static void *
smalloc(size_t nbytes)
{
	return srealloc(NULL, nbytes);
}

static void
usage(void)
{
#ifdef	SUS
	fprintf(stderr, "\
Usage: %s [user]\n\
       %s -a [user]\n\
       %s -G [-n] [user]\n\
       %s -g [-nr] [user]\n\
       %s -u [-nr] [user]\n",
		progname, progname, progname, progname, progname);
#else
	fprintf(stderr, "usage: %s [-a]\n", progname);
#endif
	exit(2);
}

static int	putspace;

static void
print_id(char *string, unsigned id, char *name)
{
	if (restriction) {
		if (putspace)
			putchar(' ');
		if (nflag)
			printf("%s", name);
		else
			printf("%u", id);
	} else {
		printf("%s%s=%u", putspace ? " " : "", string, id);
		if (name)
			printf("(%s)", name);
	}
	putspace++;
}

static void
print_supp(unsigned id, char *name)
{
	if (restriction) {
		if (putspace++)
			putchar(' ');
		if (nflag)
			printf("%s", name);
		else
			printf("%u", id);
	} else {
		static int	putcomma;

		if (putcomma++ == 0)
			printf(" groups=");
		else
			putchar(',');
		printf("%u", id);
		if (name)
			printf("(%s)", name);
	}
}

static void
supplementary(int me, char *name, gid_t mygid, gid_t myegid)
{
	struct group	*grp;
	gid_t	*groups;
	int	i, count;

	if (me) {
		if ((count = getgroups(0, NULL)) > 0) {

			groups = smalloc(count * sizeof *groups);
			getgroups(count, groups);
			for (i = 0; i < count; i++) {
				if (mygid != myegid || mygid != groups[i] ||
						aflag > 1)
				{
					grp = getgrgid(groups[i]);
					print_supp(groups[i], grp ?
						grp->gr_name : NULL);
				}
			}
			free(groups);
		}
	} else if (name) {
		setgrent();
		while ((grp = getgrent()) != NULL) {
			if (mygid != myegid || mygid != grp->gr_gid ||
					aflag > 1)
			{
				if (grp->gr_mem)
					for (i = 0; grp->gr_mem[i]; i++)
						if (strcmp(grp->gr_mem[i],
								name) == 0)
							print_supp(grp->gr_gid,
								grp->gr_name);
			}
		}
		endgrent();
	}
}

static void
id(int me, uid_t uid, uid_t euid, gid_t gid, gid_t egid)
{
	struct passwd	*pwd;
	struct group	*grp;
	char	*name;

	pwd = getpwuid(uid);
	if (restriction == 0 || (restriction == 'u' && rflag))
		print_id("uid", uid, pwd ? pwd->pw_name : NULL);
	if (pwd) {
		name = smalloc(strlen(pwd->pw_name) + 1);
		strcpy(name, pwd->pw_name);
	} else
		name = NULL;
	grp = getgrgid(gid);
	if (restriction == 0 || (restriction == 'g' && rflag) ||
			restriction == 'G')
		print_id("gid", gid, grp ? grp->gr_name : NULL);
	if ((restriction == 0 && uid != euid) ||
			(restriction == 'u' && rflag == 0)) {
		pwd = getpwuid(euid);
		print_id("euid", euid, pwd ? pwd->pw_name : NULL);
	}
	if (((restriction == 0 || restriction == 'G') && gid != egid) ||
			(restriction == 'g' && rflag == 0)) {
		grp = getgrgid(egid);
		print_id("egid", egid, grp ? grp->gr_name : NULL);
	}
	if ((restriction == 0 && aflag) || restriction == 'G')
		supplementary(me, name, gid, egid);
	putchar('\n');
}

int
main(int argc, char **argv)
{
#ifdef	SUS
	const char	optstring[] = "aGgnru";
#else
	const char	optstring[] = "a";
#endif
	int	i, me;
	uid_t	uid, euid;
	gid_t	gid, egid;

#ifdef	__GLIBC__
	putenv("POSIXLY_CORRECT=1");
#endif
	progname = basename(argv[0]);
#ifdef	SUS
	aflag = 1;
#endif
	while ((i = getopt(argc, argv, optstring)) != EOF) {
		switch (i) {
		case 'a':
			aflag = 2;
			break;
		case 'n':
			nflag = 1;
			break;
		case 'r':
			rflag = 1;
			break;
		case 'G':
		case 'g':
		case 'u':
			if (restriction)
				usage();
			restriction = i;
			break;
		default:
			usage();
		}
	}
	if (restriction == 0 && (nflag || rflag))
		usage();
	if (restriction != 0 && aflag > 1)
		usage();
	if (restriction == 'G' && rflag)
		usage();
#ifdef	SUS
	if (argc - optind == 1) {
		struct passwd	*pwd;

		if ((pwd = getpwnam(argv[optind])) == NULL) {
			fprintf(stderr, "%s: invalid user name: %s\n",
					progname, argv[optind]);
			exit(1);
		}
		me = 0;
		uid = euid = pwd->pw_uid;
		gid = egid = pwd->pw_gid;
	} else if (argc < optind > 1) {
		usage();
	} else
#endif	/* SUS */
	{
		me = 1;
		uid = getuid();
		euid = geteuid();
		gid = getgid();
		egid = getegid();
	}
	id(me, uid, euid, gid, egid);
	return 0;
}
